### PR TITLE
Randomized svd

### PR DESCRIPTION
**Pull request by tomoleary**
-Implemented accuracy enhanced SVD algorithm from https://arxiv.org/pdf/1607.01649.pdf
-Implemented unit tests for randomized eigensolver as well as randomized SVD and MultiVector orthogonalization

### CODE REVIEW

File: hippylib/algorithms/randomizedSVD.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+# See file COPYRIGHT for details.
+#
+# This file is part of the hIPPYlib library. For more information and source code
+# availability see https://hippylib.github.io.
+#
+# hIPPYlib is free software; you can redistribute it and/or modify it under the
+# terms of the GNU General Public License (as published by the Free
+# Software Foundation) version 2.0 dated June 1991.
+
+from dolfin import Vector, MPI
```
[the remark this answers is not in the document]
Comment (tomoleary) [reply hippylib/algorithms/randomizedSVD.py]: Done

File: hippylib/test/test_randomizedEigensolver.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        alpha = dl.Constant(1.0)
+        varfA = dl.inner(dl.grad(uh), dl.grad(vh))*dl.dx +\
+                    alpha*dl.inner(uh,vh)*dl.dx
+        A = dl.assemble(varfA)
+        Asolver = dl.PETScKrylovSolver(self.A.mpi_comm(), "cg", amg_method())
+        Asolver.set_operator(self.A)
+        Asolver.parameters["maximum_iterations"] = 100
+        Asolver.parameters["relative_tolerance"] = 1e-12
+
+        ## Set up M
```
[the remark this answers is not in the document]
Comment (tomoleary) [reply hippylib/test/test_randomizedEigensolver.py]: Done

File: hippylib/test/test_randomizedEigensolver.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+from numpy.testing import assert_allclose
+
+import sys
+sys.path.append('../../')
+from hippylib import (MultiVector, Random, assemblePointwiseObservation,
+                      singlePass, singlePassG, doublePass, doublePassG, MatMvMult)
+                      
+
+
+class Aop:
```
[the remark this answers is not in the document]
Comment (tomoleary) [reply hippylib/test/test_randomizedEigensolver.py]: Done

**Comment by joshuawchen** [comment]
I can look at it thoroughly on Monday or Tuesday.

File: hippylib/algorithms/randomizedSVD.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    - :code:`sigma`: the estimate of the :math:`k` dominant singular values of :math:`A`.
+    - :code:`V`: the estimate of the :math:`k` dominant right singular vectors of :math:`A,\\, V^T V = I_k`.
+    
+    """
+
+    # Check compatibility of operator A
+    assert hasattr(A,'transpmult'), 'Operator A must have member function transpmult'
+
+    nvec  = Omega.nvec()
+    assert(nvec >= k )
```
Comment (AIlona) [inline hippylib/algorithms/randomizedSVD.py]: Not a big deal at all, but maybe you should get rid of parentheses (and add a message)  for consistency purposes.

File: hippylib/algorithms/randomizedSVD.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    - :code:`sigma`: the estimate of the :math:`k` dominant singular values of :math:`A`.
+    - :code:`V`: the estimate of the :math:`k` dominant right singular vectors of :math:`A,\\, V^T V = I_k`.
+    
+    """
+
+    # Check compatibility of operator A
+    assert hasattr(A,'transpmult'), 'Operator A must have member function transpmult'
+
+    nvec  = Omega_c.nvec()
+    assert(nvec >= k )
```
Comment (AIlona) [inline hippylib/algorithms/randomizedSVD.py]: Same comment about parentheses and message.

**Pull request merged**